### PR TITLE
fix(dep update): helm dep update is not respecting the "version" stip…

### LINKED ISSUE #9348: Helm dep update is not respecting the "version" stipulated in the requirements.
Helm dep update is not respecting the "version" stipulated in the parent requirements file. 

Parent chart's requirement.yaml example:
```
- name: child-chart
  version: 1.0.1
  repository: "file://../child-chart"
  condition: child-chart.enabled
```

Child chart's chart.yaml example:
```
apiVersion: v1
description: child-chart
name: child-chart
version: 1.0.0
appVersion: "1.0"
```

When running `helm dep up`, the parent chart will include the child chart in the charts folder, i.e.

parent/charts/child-charts-1.0.0.tgz <-- This should not happen, and the `helm dep up` should respect the version stipulated in the requirements.yaml `v1.0.1` and thus should fail. When using helm v2.17.0, and helm v3.3.0, the `helm dep up` would fail in this manner.

When running `helm install parent`, the Installation will excluding the missing child-chart as there is no `parent/charts/child-charts-1.0.1.tgz` available. There is no way to easily identify if there is a missing dependency as the installation is "successful" even though one of the dependencies are missing.

Output of `helm version`:

Experienced this issues originally with v3.4.2:
```
version.BuildInfo{Version:"v3.4.2", GitCommit:"23dd3af5e19a02d4f4baa5b2f242645a1a3af629", GitTreeState:"dirty", GoVersion:"go1.15.5"}
```

Upgraded to latest stable via Brew and experienced same issue:
```
version.BuildInfo{Version:"v3.5.2", GitCommit:"167aac70832d3a384f65f9745335e9fb40169dc2", GitTreeState:"dirty", GoVersion:"go1.15.7"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T18:49:28Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.6", GitCommit:"b1d75deca493a24a2f87eb1efde1a569e52fc8d9", GitTreeState:"clean", BuildDate:"2018-12-16T04:30:10Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
Rancher v2.4.4

### PR DESCRIPTION
**Pull request by cndoit18**
Signed-off-by: cndoit18 <cndoit18@outlook.com>

**What this PR does / why we need it**:
fixes #9348 

**Special notes for your reviewer**:

**If applicable**:
- [x] this PR contains unit tests
- [x] this PR has been tested for backward compatibility

### CODE REVIEW

File: internal/resolver/resolver.go [hunk 1]
```diff
@@ -84,6 +89,17 @@ func (r *Resolver) Resolve(reqs []*chart.Dependency, repoNames map[string]string
 				return nil, err
 			}
 
+			v, err := semver.NewVersion(ch.Metadata.Version)
```
[the remark this answers is not in the document]
Comment (cndoit18) [reply internal/resolver/resolver.go:90]: > There is a comment a few lines before this that needs to be removed. It reads..
> 
> ```
> // The version of the chart locked will be the version of the chart
> // currently listed in the file system within the chart.
> ```
> 
> Can you please remove it?

ok, it is done

File: internal/resolver/resolver.go [hunk 2]
```diff
@@ -57,6 +57,11 @@ func (r *Resolver) Resolve(reqs []*chart.Dependency, repoNames map[string]string
 	locked := make([]*chart.Dependency, len(reqs))
 	missing := []string{}
 	for i, d := range reqs {
+		constraint, err := semver.NewConstraint(d.Version)
+		if err != nil {
+			return nil, errors.Wrapf(err, "dependency %q has an invalid version/constraint format", d.Name)
+		}
+
 		if d.Repository == "" {
 			// Local chart subfolder
```
Comment (mattfarina) [inline internal/resolver/resolver.go:66]: I just realized that this doesn't fix #9348. In #9348 the repository exists as it starts with `file://`. The modified code path here is for the case where there is no repository set. Instead, this code touches on the code path for charts who exist only in the `charts/` sub-directory of a chart. There is no outside location as no repository is set.

Can this PR be updated to fix #9348, please.

**Review by mattfarina — CHANGES_REQUESTED** [review]

File: internal/resolver/resolver.go (hunk 2 above)
Comment (cndoit18) [reply internal/resolver/resolver.go:66]: Sorry, I didn't quite get it, the modified code path here is for the case where that the repository exists.

See, line: 73

**Review by technosophos — APPROVED** [review]
I am not sure whether it covers #9348, but as far as fixing a bug for known repos, this does seem to fix a bug.

File: internal/resolver/resolver.go (hunk 2 above)
Comment (mattfarina) [reply internal/resolver/resolver.go:66]: Actually, it was my fault. It does resolve the issue. I was reading it wrong in the GitHub interface. Some code was collapsed. It's good.

**Review by mattfarina — APPROVED** [review]

**Comment by tamalsaha** [comment]
Hello, this pr has introduced a regression in v3.6.3. Bug report: https://github.com/helm/helm/issues/9944

**Pull request merged**